schemas/context/namespace.schema.json

```diff
@@ -0,0 +1,37 @@
+{
+   "meta:license":[
+      "Copyright 2017 Adobe Systems Incorporated. All rights reserved.",
+      "This work is licensed under a Creative Commons Attribution 4.0 International (CC BY 4.0) license",
+      "you may not use this file except in compliance with the License. You may obtain a copy",
+      "of the License at https://creativecommons.org/licenses/by/4.0/"
+   ],
+   "$id":"https://ns.adobe.com/xdm/context/namespace",
+   "$schema":"http://json-schema.org/draft-06/schema#",
+   "title":"Namespace",
+   "type":"object",
+   "description":"The Datasource acts as a namespace or unique identifier associated with a collection of data. Each EndUserID is associated with a given Datasource. A user can create and obtain information about a Datasource during the solution onboarding.\n",
+   "definitions":{
+      "namespace":{
+         "properties":{
+            "xdm:id":{
+               "title":"Identifier",
+               "type":"integer",
+               "description":"The numeric ID associated with this Datasource. This would be provided by the individual or system that created the Datasource.\n"
+            },
+            "xdm:code":{
```

Comment:
I'm completely at a loss why we need both and what the difference is. Above it seems to make some reference to Adobe Audience Manager, which seems to be an implementation detail.

Reply:
Paging @cdegroot-adobe ;)

Reply:
I think we should be clear which properties are core and which are implementation detail (and can be delegated to an extension). Giving the datasource an identifier (ideally an uniform identifier of the resource) makes a lot of sense. Forcing in an additional identifier that is only relevant to a soon-to-be legacy system will only lead to confusion.

You can count my confusion as a small taste of what's to come when we expose this to users with less understanding of our systems than I have.

Reply:
The systems we are using here are certainly not "soon-to-be legacy", they are being re-organized to provide a single consistent service for customers/partners/adobe etc. Nothing is going away.
What we are doing here does need to be expanded to cater for other "governors or data" services beyond Adobe, and that would certainly be a URI. But the values that need to be processed for a namespace and identity itself must be defined an a way that also meets performance goals for implementations. The way Identity works is not for the faint of heart.

Another aspect of this duality that we need to outline more is the ingest/egress vs enriched processing. Systems generating ExperienceEvents or Profiles will very often not be able to easily generate the XIDs, so will provide only the Id and namespace. Systems that are XID aware  will use and generate XIDs and be able to use them for super low latency processing between each other, especially when identity graphs are in play. XDM is also for this XID-aware system to system interoperability, not just ingress/egress. If the ingress/egress system does populate the XIDs then they can expect that those XIDs will be retained in responses.

Reply:
In my mind, both `xdm:id` and `xdm:code` can be combined into a simple `@id` that follows a pattern like `https://data.adobe.io/{namespace|datasource|…}/<code>/<numeric_id>`. This gives you the ease of processing it in systems that can read only the code or the numeric ID today, but won't lock you in in the future when you decide to use only codes, only numbers, or a completely different identification system.

Reply:
As a representation yes. How much that is helping human readability is open in my mind. Considering the primary processing systems in our industry are SQL, R, Scala etc this is not that effective. It would be problematic to natively use it, thus complex processing is needed Unicode parsing, secure URL parsing etc. Scale this to trillions of parsings a day with sub ms SLAs and it is not realistic.

Reply:
This actually isn't true.  The identifier is not a composite, like <id + code> or <code + id>.  Id is the identifier, and managed by the system.  Code is an alias, and managed by the application (in the context of a tenant).  There is an expected data flow where a user includes only the code, and the system (knowing the tenant context) looks up the ID.  The ID (which is global across tenants) is then used in the rest of the flow.

Reply:
So, do we need both? It sounds like `id` is globally unique, but `code` only works when you know the tenant, which we haven't even modeled.

Reply:
Add to that that `id` is required, but `code` is optional. For the flow that @kstreeter describes it sounds like `id` is optional, but `code` is required.

Reply:
You need one, or the other, or both.  ID is not necessarily globally unique: it is scoped to the namespace.

The code is actually scoped to the namespace, too.  My description above was a little incomplete there: the code is scoped to a namespace, which may internally be owned by a tenant but that is probably an implementation detail at this level.

Reply:
Reading through this, I'm not 100% sure what the resolution is. @trieloff  - What open items do you still see?

Reply:
Resolution:
- [ ] update description: explain that the `code` is a shortcut to the full `@id` and that either `code` or `@id` can be used
- [ ] update schema with an `oneOf/required` construct, so that validation enforces presence of either `code` or `@id`
- [ ] provide examples and invalid examples for each

/cc @cmathis @jwen-adobe @kstreeter

Reply:
@trieloff @cmathis @kstreeter Done. I did not use oneOf here as comment from Kevin stated both can be used as well. I replaced "required" with "minProperties" though.